File: lib/test/t_CloudMesher_std.cxx
#include <iostream>
#include <cmath>

#include <openturns/OT.hxx>
#include <openturns/OTtestcode.hxx>

#include "otmeshing/otmeshing.hxx"

using namespace OT;
using namespace OT::Test;
using namespace OTMESHING;

int main()
{
  // Test default constructor and string representation
  CloudMesher mesher;
  std::cout << "mesher=" << mesher << std::endl;
  assert_equal(mesher.getClassName(), String("CloudMesher"));

  // Test 2D triangulation with BASIC (default)
  {
    Sample points(5, 2);
    points[0] = Point({3.0, 0.0});
    points[1] = Point({2.0, 0.0});
    points[2] = Point({2.0, 0.75});
    points[3] = Point({2.5, 0.75});
    points[4] = Point({3.0, 0.2});
    Mesh mesh = mesher.build(points);
    assert_equal(mesh.getDimension(), 2ul);
    assert_equal(mesh.getSimplices().getSize(), 3ul);
    assert_equal(mesh.isValid(), true);
    assert_equal(mesh.isConvex(), true);
    assert_almost_equal(mesh.getVolume(), 0.6125, 1e-12, 1e-12);
  }

  // Test 2D triangulation with DELAUNAY
  {
    CloudMesher delMesher(CloudMesher::DELAUNAY);
    Sample points(5, 2);
    points[0] = Point({3.0, 0.0});
    points[1] = Point({2.0, 0.0});
    points[2] = Point({2.0, 0.75});
    points[3] = Point({2.5, 0.75});
    points[4] = Point({3.0, 0.2});
    Mesh mesh = delMesher.build(points);
    assert_equal(mesh.getDimension(), 2ul);
    assert_equal(mesh.getSimplices().getSize(), 3ul);
    assert_equal(mesh.isValid(), true);
    assert_equal(mesh.isConvex(), true);
    assert_almost_equal(mesh.getVolume(), 0.6125, 1e-12, 1e-12);
  }

  // Test 1D case: only min and max become vertices
  {
    Sample points(3, 1);
    points[0] = Point({2.5});
    points[1] = Point({1.5});
    points[2] = Point({3.0});
    Mesh mesh = mesher.build(points);
    assert_equal(mesh.getDimension(), 1ul);
    assert_equal(mesh.isValid(), true);
    assert_equal(mesh.isConvex(), true);
    assert_almost_equal(mesh.getVolume(), 1.5, 1e-12, 1e-12);
    assert_equal(mesh.getVertices().getSize(), 2ul);
    assert_equal(mesh.getSimplices().getSize(), 1ul);
    assert_almost_equal(mesh.getVertices()[0][0], 1.5, 1e-12, 1e-12);
    assert_almost_equal(mesh.getVertices()[1][0], 3.0, 1e-12, 1e-12);
  }

  // Test unit hypercube triangulations
  {
    for (UnsignedInteger method = 0; method <= 1; ++method)
    {
      CloudMesher cubeMesher(static_cast<CloudMesher::TriangulationMethod>(method));
      for (UnsignedInteger dim = 1; dim <= 6; ++dim)
      {
        // Generate unit hypercube vertices
        const UnsignedInteger nbVertices = static_cast<UnsignedInteger>(std::pow(2.0, static_cast<double>(dim)));
        Sample vertices(nbVertices, dim);
        for (UnsignedInteger i = 0; i < nbVertices; ++i)
        {
          for (UnsignedInteger j = 0; j < dim; ++j)
          {
            vertices[i][j] = (i >> j) & 1;
          }
        }
        Mesh mesh = cubeMesher.build(vertices);
        assert_equal(mesh.getDimension(), dim);
        assert_equal(mesh.getVertices().getSize(), vertices.getSize());
        assert_equal(mesh.isValid(), true);
        assert_equal(mesh.isConvex(), true);
        assert_almost_equal(mesh.getVolume(), 1.0, 1e-12, 1e-12);
      }
    }
  }

  // Error cases: null dimension
  {
    Sample empty;
    bool caught = false;
    try
    {
      mesher.build(empty);
    }
    catch (const InvalidArgumentException &)
    {
      caught = true;
    }
    assert_equal(caught, true);
  }

  // Error cases: insufficient points for 2D
  {
    Sample tooFew(1, 2);
    tooFew[0] = Point({1.0, 2.0});
    bool caught = false;
    try
    {
      mesher.build(tooFew);
    }
    catch (const InvalidArgumentException &)
    {
      caught = true;
    }
    assert_equal(caught, true);
  }

  // Error cases: 1D with single point
  {
    Sample tooFew(1, 1);
    tooFew[0] = Point({1.0});
    bool caught = false;
    try
    {
      mesher.build(tooFew);
    }
    catch (const InvalidArgumentException &)
    {
      caught = true;
    }
    assert_equal(caught, true);
  }

  std::cout << "All C++ tests passed!" << std::endl;
  return 0;
}

File: python/test/t_CloudMesher_std.py
#! /usr/bin/env python

import math
import openturns as ot
import openturns.testing as ott
import otmeshing

ot.TESTPREAMBLE()

# basic 2d triangulation - BASIC method
mesher = otmeshing.CloudMesher()
print("mesher=", mesher)
assert "CloudMesher" in repr(mesher)
assert mesher.getClassName() == "CloudMesher"

vertices = [[3.0, 0.0], [2.0, 0.0], [2.0, 0.75], [2.5, 0.75], [3.0, 0.2]]
triangulation = mesher.build(vertices)
vol = triangulation.getVolume()
print(f"-- 2d BASIC={repr(triangulation)} vol={vol}")
assert triangulation.getDimension() == 2
assert len(triangulation.getSimplices()) == 3
assert triangulation.isValid()
ott.assert_almost_equal(vol, 0.6125)
assert triangulation.isConvex()
# check all vertices are part of the input
verts = [[v[0], v[1]] for v in triangulation.getVertices()]
for v in verts:
    assert v in vertices, f"vertex {v} not in input"

# basic 2d triangulation - DELAUNAY method
mesher = otmeshing.CloudMesher(otmeshing.CloudMesher.DELAUNAY)
triangulation = mesher.build(vertices)
vol = triangulation.getVolume()
print(f"-- 2d DELAUNAY={repr(triangulation)} vol={vol}")
assert triangulation.getDimension() == 2
assert len(triangulation.getSimplices()) == 3
assert triangulation.isValid()
ott.assert_almost_equal(vol, 0.6125)
assert triangulation.isConvex()

# 1D case
mesher = otmeshing.CloudMesher()
triangulation = mesher.build([[2.5], [1.5], [3.0]])
assert triangulation.getDimension() == 1
assert triangulation.isValid()
assert triangulation.isConvex()
ott.assert_almost_equal(triangulation.getVolume(), 1.5)
# 1D special case: only min and max become vertices
assert len(triangulation.getVertices()) == 2
assert len(triangulation.getSimplices()) == 1
verts = triangulation.getVertices()
ott.assert_almost_equal(verts[0], [1.5])
ott.assert_almost_equal(verts[1], [3.0])

# nd triangulation of the unit hypercube
for method in [otmeshing.CloudMesher.BASIC, otmeshing.CloudMesher.DELAUNAY]:
    mesher = otmeshing.CloudMesher(method)
    for dim in range(1, 7):
        print(f"-- cube dim={dim} method={method}")
        vertices = ot.Box([0] * dim).generate()
        triangulation = mesher.build(vertices)
        assert triangulation.getDimension() == dim
        vol = triangulation.getVolume()
        print(f"vol={vol} triangulation={repr(triangulation)[:2000]}")
        assert len(triangulation.getVertices()) == len(vertices)
        assert triangulation.isValid()
        ott.assert_almost_equal(vol, 1.0)
        assert triangulation.isConvex()

# nd triangulation of the unit hypersphere
for method in [otmeshing.CloudMesher.BASIC, otmeshing.CloudMesher.DELAUNAY]:
    mesher = otmeshing.CloudMesher(method)
    for dim in range(1, 5):
        print(f"-- sphere dim={dim} method={method}")
        vertices = ot.Normal(dim).getSample(1000)
        for i in range(len(vertices)):
            vI = vertices[i]
            vertices[i] = vI / vI.norm()
        triangulation = mesher.build(vertices)
        assert triangulation.getDimension() == dim
        vol = triangulation.getVolume()
        print(f"vol={vol} triangulation={repr(triangulation)[:2000]}")
        if dim > 1:
            assert len(triangulation.getVertices()) == len(vertices)
        assert triangulation.isValid()
        vol_ref = math.pi ** (dim / 2) / math.gamma(dim / 2 + 1)
        ott.assert_almost_equal(vol, vol_ref, 0.1, 0.0)
        assert triangulation.isConvex()

# error cases: null dimension
with ott.assert_raises(TypeError):
    otmeshing.CloudMesher().build([])

# error cases: invalid empty point
with ott.assert_raises(TypeError):
    otmeshing.CloudMesher().build([[]])

# error cases: insufficient points for dimension
with ott.assert_raises(TypeError):
    otmeshing.CloudMesher().build([[1.0]])
with ott.assert_raises(TypeError):
    otmeshing.CloudMesher().build([[1.0, 2.0]])
